### PR TITLE
Added unittests for EditDepartmentCommand and EditDepartmentRequestValidator

### PR DESCRIPTION
**Pull request by Vasis3038**
https://trello.com/c/k9CGEhUy/119-departmentservice-add-unittests-for-edit

### CODE REVIEW

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    _autoMocker.GetMock<IPatchDbDepartmentMapper>().Reset();
+    _autoMocker.GetMock<IDepartmentRepository>().Reset();
+
+    _autoMocker
+      .Setup<IAccessValidator, Task<bool>>(x => x.HasRightsAsync(Rights.AddEditRemoveDepartments))
+      .ReturnsAsync(true);
+
+    _autoMocker
+      .Setup<IEditDepartmentRequestValidator, bool>(x => x.ValidateAsync(It.IsAny<ValueTuple<Guid, JsonPatchDocument<EditDepartmentRequest>>>(), default).Result.IsValid)
+      .Returns(true);
```
Comment (maybee16) [inline test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs:146]: ReturnsAsync

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+    _autoMocker
+      .Setup<IPatchDbDepartmentMapper, JsonPatchDocument<DbDepartment>>(x => x.Map(_request))
+      .Returns(_dbRequest);
+
+    _autoMocker
+      .Setup<IDepartmentRepository, Task<bool>>(x => x.EditAsync(It.IsAny<Guid>(), It.IsAny<JsonPatchDocument<DbDepartment>>()))
+      .Returns(Task.FromResult(true));
```
Comment (maybee16) [inline test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs:154]: ReturnsAsync

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+    _autoMocker.Verify<IDepartmentRepository>(
+      x => x.EditAsync(It.IsAny<Guid>(), It.IsAny<JsonPatchDocument<DbDepartment>>()),
+      Times.Never);
+  }
+
+  [Test]
+  public async Task ShouldReturnFailedResponseWhenValidationIsFailedAsync()
+  {
+
```
Comment (maybee16) [inline test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs:191]: Empty line

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    _autoMocker
+      .Setup<IEditDepartmentRequestValidator, bool>(x => x.ValidateAsync(It.IsAny<ValueTuple<Guid, JsonPatchDocument<EditDepartmentRequest>>>(), default).Result.IsValid)
+      .Returns(false);
```
Comment (maybee16) [inline test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs:194]: ReturnsAsync

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs (hunk 2 above)
Comment (Vasis3038) [reply test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs]: Done

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs (hunk 3 above)
Comment (Vasis3038) [reply test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs]: Done

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs (hunk 4 above)
Comment (Vasis3038) [reply test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs]: Done

File: test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+    Assert.True(_validator.ValidateAsync((_departmentId, _request)).Result.IsValid);
+  }
+
+  [Test]
+  public void ShouldThrowExceptionWhenCategoryIdDoesNotExist()
+  {
+    _autoMocker
+      .Setup<ICategoryRepository, Task<bool>>(x => x.IdExistAsync(It.IsAny<Guid>()))
+      .Returns(Task.FromResult(false));
```
Comment (AnnPostnikova) [inline test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs:184]: ReturnsAsync

File: test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+  [SetUp]
+  public void SetUp()
+  {
+    _autoMocker.GetMock<IDepartmentRepository>().Reset();
+    _autoMocker.GetMock<ICategoryRepository>().Reset();
+    _autoMocker.GetMock<IEditDepartmentRequestValidator>().Reset();
+
+    _autoMocker
+      .Setup<IDepartmentRepository, Task<bool>>(x => x.NameExistAsync(It.IsAny<String>(), It.IsAny<Guid>()))
+      .Returns(Task.FromResult(false));
```
Comment (AnnPostnikova) [inline test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs:44]: ReturnsAsync

File: test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+    _autoMocker
+      .Setup<IDepartmentRepository, Task<bool>>(x => x.ShortNameExistAsync(It.IsAny<String>(), It.IsAny<Guid>()))
+      .Returns(Task.FromResult(false));
```
Comment (AnnPostnikova) [inline test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs:48]: Same

File: test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+    _autoMocker
+      .Setup<ICategoryRepository, Task<bool>>(x => x.IdExistAsync(It.IsAny<Guid>()))
+      .Returns(Task.FromResult(true));
```
Comment (AnnPostnikova) [inline test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs:52]: Same

File: test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+    Assert.False(_validator.ValidateAsync((_departmentId, _request)).Result.IsValid);
+  }
+
+  [Test]
+  public void ShouldThrowExceptionWhenNameAlreadyExists()
+  {
+    _autoMocker
+      .Setup<IDepartmentRepository, Task<bool>>(x => x.NameExistAsync(It.IsAny<String>(), It.IsAny<Guid>()))
+      .Returns(Task.FromResult(true));
```
Comment (AnnPostnikova) [inline test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs:124]: same

File: test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+    Assert.False(_validator.ValidateAsync((_departmentId, _request)).Result.IsValid);
+  }
+
+  [Test]
+  public void ShouldThrowExceptionWhenShortNameAlreadyExists()
+  {
+    _autoMocker
+      .Setup<IDepartmentRepository, Task<bool>>(x => x.ShortNameExistAsync(It.IsAny<String>(), It.IsAny<Guid>()))
+      .Returns(Task.FromResult(true));
```
Comment (AnnPostnikova) [inline test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs:158]: same

Comment (Vasis3038) [reply test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs]: Done

File: test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs (hunk 7 above)
Comment (Vasis3038) [reply test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs]: Done

File: test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs (hunk 6 above)
Comment (Vasis3038) [reply test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs]: Done

File: test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs (hunk 5 above)
Comment (Vasis3038) [reply test/DepartmentService.Validation.UnitTests/EditDepartmentRequestValidatorTest.cs]: Done

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+  public async Task ShouldEditDepartmentSuccesfullAsync()
+  {
+    OperationResultResponse<bool> expectedResponse = new()
+    {
+      Body = true
+    };
+
+    SerializerAssert.AreEqual(expectedResponse, await _command.ExecuteAsync(Guid.NewGuid(), _request));
+
+    _autoMocker.Verify<IAccessValidator>(
```
Comment (VLiamin) [inline test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs:229]: May be add Verify method? Like in others services.

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+      x => x.Map(It.IsAny<JsonPatchDocument<EditDepartmentRequest>>()),
+      Times.Never);
+
+    _autoMocker.Verify<IDepartmentRepository>(
+      x => x.EditAsync(It.IsAny<Guid>(), It.IsAny<JsonPatchDocument<DbDepartment>>()),
+      Times.Never);
+  }
+
+  [Test]
+  public async Task ShouldEditDepartmentSuccesfullAsync()
```
Comment (AnnPostnikova) [inline test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs:220]: better EditDepartmentSuccesfullTestAsync

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+      .Setup<IPatchDbDepartmentMapper, JsonPatchDocument<DbDepartment>>(x => x.Map(_request))
+      .Returns(_dbRequest);
+
+    _autoMocker
+      .Setup<IDepartmentRepository, Task<bool>>(x => x.EditAsync(It.IsAny<Guid>(), It.IsAny<JsonPatchDocument<DbDepartment>>()))
+      .ReturnsAsync(true);
+  }
+
+  [Test]
+  public async Task ShouldReturnFailedResponseWhenUserHasNotRightAsync()
```
Comment (AnnPostnikova) [inline test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs:158]: better NotEnoughRightTestAsync

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+      x => x.Map(It.IsAny<JsonPatchDocument<EditDepartmentRequest>>()),
+      Times.Never);
+
+    _autoMocker.Verify<IDepartmentRepository>(
+      x => x.EditAsync(It.IsAny<Guid>(), It.IsAny<JsonPatchDocument<DbDepartment>>()),
+      Times.Never);
+  }
+
+  [Test]
+  public async Task ShouldReturnFailedResponseWhenValidationIsFailedAsync()
```
[the remark this answers is not in the document]
Comment (Vasis3038) [reply test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs]: Done

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs (hunk 13 above)
Comment (Vasis3038) [reply test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs]: Done

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs (hunk 12 above)
Comment (Vasis3038) [reply test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs]: Done

File: test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs (hunk 11 above)
Comment (Vasis3038) [reply test/DepartmentService.Business.UnitTests/Commands/Department/EditDepartmentCommandTests.cs]: Done

**Review by AnnPostnikova — APPROVED** [review]

**Review by vorobevamv — APPROVED** [review]

**Pull request merged**